src/subcommands/mod.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright 2020 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Subcommand implementations. Interacts with parsec-client-rust.
+
+pub mod common;
+pub mod list_opcodes;
+pub mod list_providers;
+pub mod ping;
+
+use crate::cli::ParsecToolApp;
+use crate::error::ParsecToolError;
+use crate::subcommands::{
+    list_opcodes::ListOpcodesSubcommand, list_providers::ListProvidersSubcommand,
+    ping::PingSubcommand,
+};
+use anyhow::Result;
+use parsec_interface::operations::NativeOperation;
+use std::convert::TryInto;
+use structopt::StructOpt;
+
+/// A trait to represent a `parsec-tool` subcommand. Subcommands have three important properties:
+/// - They have their own command-line interface, hence the dependency on `StructOpt`.
+/// - They are convertible to a `NativeOperation` -- i.e. they can all be converted to messages to
+///   the Parsec service. The conversion is fallible.
+/// - They implement `run`, which executes the subcommand.
+pub trait ParsecToolSubcommand: StructOpt + TryInto<NativeOperation> {
+    /// Run the subcommand.
+    fn run(self, matches: ParsecToolApp) -> Result<(), ParsecToolError>;
+}
+
+/// All of the commands supported by parsec-tool. Anything added into this enum will automatically
+/// be added as a subcommand to the CLI (see src/cli.rs).
+#[derive(Copy, Clone, Debug, StructOpt)]
+pub enum Subcommand {
+    /// Pings the Parsec service and prints the wire protocol version.
+    Ping(PingSubcommand),
+
+    /// Lists the available providers supported by the Parsec service.
+    ListProviders(ListProvidersSubcommand),
+
+    /// Lists the supported opcodes for a given provider.
+    ListOpcodes(ListOpcodesSubcommand),
+}
+
+/// Runs the appropriate command given `matches`, which corresponds to the command-line invocation
+/// from the user.
+pub fn dispatch_subcommand(matches: ParsecToolApp) -> Result<(), ParsecToolError> {
```

Comment:
Similar to the previous comment, this can be implemented as a method on `Subcommand`, or maybe even on `ParsecToolApp`, depending on how you want to call it in `main()`. 

We've generally avoided having "hanging" functions if we can implement them on some type where it makes sense for them to be attached. Sometimes that's not possible, though, either because the types are external so we literally cannot implement on them, or because we need multiple implementations for different things, so we then use functions.